### DISCUSSION TITLE
[Android] Usage of gl_FragDepth with version 1.12.1 and above

### DISCUSSION THREAD

**Discussion by RGregat** [google/filament#4729] [Q&A] [answered]
Hey, I'm currently working on an [update for Sceneform Maintained](https://github.com/ThomasGorisse/sceneform-android-sdk/pull/179) to include the new Filament version **1.12.8** and stumbled over an issue regarding our Occlusion Material. 

```
material {
    name : depth,
    shadingModel : unlit,
    blending : opaque,
    vertexDomain : device,
    parameters : [
        {
            type : samplerExternal,
            name : cameraTexture
        },
        {
            type : sampler2d,
            name : depthTexture
        },
        {
            type : float4x4,
            name : uvTransform
        }
    ],
    requires : [
        uv0
    ]
}

fragment {
    void material(inout MaterialInputs material) {
        prepareMaterial(material);
        material.baseColor.rgb = inverseTonemapSRGB(texture(materialParams_cameraTexture, getUV0()).rgb);
        vec2 packed_depth = texture(materialParams_depthTexture, getUV0()).xy;
        float depth_mm = dot(packed_depth, vec2(255.f, 256.f * 255.f));
        vec4 view = mulMat4x4Float3(getClipFromViewMatrix(), vec3(0.f, 0.f, -depth_mm / 1000.f));
        float ndc_depth = view.z / view.w;
        gl_FragDepth = 1.f - ((ndc_depth + 1.f) / 2.f);
    }
}

vertex {
    void materialVertex(inout MaterialVertexInputs material) {
        material.uv0 = mulMat4x4Float3(materialParams.uvTransform, vec3(material.uv0.x, material.uv0.y, 0.f)).xy;
    }
}
```

This material worked fine with Version **1.12.0** but stopped working with version **1.12.1** and above. 

![DepthIssueFilament1 12 1+](https://user-images.githubusercontent.com/3974162/137076368-95e74cc0-aba0-4628-a94a-97fead1ab4ee.jpg)

So my assumption is that the usage of gl_FragDepth has changed somehow or something else.. I have read through the release notes but didn't found any crucial change. Maybe someone has an idea what happened between 1.12.0 and 1.12.1

**Comment by romainguy** [comment]
Try to write the inverse depth (remove your 1.0 - when writing to FragDepth). @pixelflinger can help you more.

**Reply by RGregat** [reply]
If I write `gl_FragDepth = ((ndc_depth + 1.f) / 2.f);` the Panda is completly gone.

**Comment by pixelflinger** [comment, marked as answer]
We recently changed the projection matrix so it uses a clip space between 1 and 0 for the depth (1=near, 0=infinity).
When running on vk and metal, the actual clip space is set to that. But in gl, it depends on whether the clip_control extension is there. If it's not there, we're correcting the projection matrix clip space in the vertex shader.

Bottom line is, the depth value in the frame buffer should be between 1 (near) and 0. And, if you're using our projection matrix, you need to know that its clip space for Z is 1 to 0. Meaning that you do not have to rescale the depth value between -1 and 1, when you need the clip space.

**Reply by RGregat** [reply]
So `gl_FragDepth = ndc_depth` should be working. 
A first test was successful. I have to do some more tests but this looks now promising, Thank you for your quick response @romainguy and @pixelflinger
